core/src/main/scala/akka/kafka/internal/ConsumerStage.scala

```diff
@@ -0,0 +1,267 @@
+            require(isAvailable(out))
+            buffer = records.iterator().asScala.map { record =>
+              val offset = Consumer.Offset(
+                topic = record.topic,
+                partition = record.partition,
+                offset = record.offset
+              )
+              val committable = CommittableOffsetImpl(offset)(this)
+              Consumer.CommittableMessage(record.key, record.value, committable)
+            }
+            pushMsg(buffer.next())
+          }
+        }
+
+        def pollTimeout = {
+          val data = Option(POLL_DATA_TIMEOUT).filter(_ => isAvailable(out) && !buffer.hasNext)
+          def commit = Option(POLL_COMMIT_TIMEOUT).filter(_ => awaitingConfirmation > 0)
+          data.orElse(commit)
+        }
+
+        pollTimeout match {
+          case Some(timeout) =>
+            setupConsumer()
+            // TODO poll is blocking, can't we do this without blocking?
+            val msgs = consumer.poll(timeout.toMillis)
+            handleResult(msgs)
+            if (pollTimeout.isDefined) schedulePoll()
+          case None =>
+        }
+
+        if (stopping && !isClosed(out)) {
+          logger.debug("Stop producing messages")
+          complete(out)
+        }
+
+        if (stopping && awaitingConfirmation == 0) {
+          completeStage()
+        }
+      }
+
+      val pollCallback = getAsyncCallback[Unit] { _ => schedulePoll() }
+
+      private def pushMsg(msg: Consumer.CommittableMessage[K, V]): Unit = {
+        logger.trace("Push element {}", msg)
+        push(out, msg)
+      }
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          if (!buffer.hasNext) poll()
+          else pushMsg(buffer.next())
+        }
+      })
+
+      private val decrementConfirmation = getAsyncCallback[Unit] { _ =>
+        awaitingConfirmation -= 1
+        logger.trace(s"Commits in progress {}", awaitingConfirmation.toString)
+      }
+
+      // impl of GraphStage.Committer that is called from the outside via CommittableOffsetImpl
+      override def commit(offset: Consumer.Offset): Future[Done] = {
+        // committed offset should be the next message the application will consume, i.e. +1
+        commit(Map(new TopicPartition(offset.topic, offset.partition) ->
+          new OffsetAndMetadata(offset.offset + 1)))
+      }
+
+      // impl of GraphStage.Committer that is called from the outside via CommittableOffsetBatchImpl
+      override def commit(offsets: Map[TopicPartition, OffsetAndMetadata]): Future[Done] = {
+        // not important that we complete with failure during stopping race,
+        // it will be completed with TimeoutException
+        if (stoppedPromise.isCompleted)
+          Future.failed(new IllegalStateException("ConsumerStage is stopped, offset commit not performed"))
+        else {
+          val done = Promise[Done]()
+          val c = materializer.scheduleOnce(COMMIT_TIMEOUT, new Runnable {
```

Comment:
yes, this is called from the outside, bummer.
The intention was to schedule a timeout to fail the future to avoid uncompleted futures. It could be a race between stopping and `commitCallback.invoke`. If stopped `invoke` will do nothing.